include/nbl/video/IDescriptorPool.h

```diff
@@ -29,12 +33,156 @@ class NBL_API IDescriptorPool : public core::IReferenceCounted, public IBackendO
+            m_combinedImageSamplerStorage = std::make_unique<StorageTrivializer<SCombinedImageSampler>[]>(m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER]);
+            m_storageImageStorage = std::make_unique<StorageTrivializer<core::smart_refctd_ptr<IGPUImageView>>[]>(m_maxDescriptorCount[asset::EDT_STORAGE_IMAGE] + m_maxDescriptorCount[asset::EDT_INPUT_ATTACHMENT]);
+            m_UBO_SSBOStorage = std::make_unique<StorageTrivializer<core::smart_refctd_ptr<IGPUBuffer>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER] + m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER_DYNAMIC] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
+            m_UTB_STBStorage = std::make_unique<StorageTrivializer<core::smart_refctd_ptr<IGPUBufferView>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_TEXEL_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
+            m_accelerationStructureStorage = std::make_unique<StorageTrivializer<core::smart_refctd_ptr<IGPUAccelerationStructure>>[]>(m_maxDescriptorCount[asset::EDT_ACCELERATION_STRUCTURE]);
+        }
+
+        ~IDescriptorPool()
+        {
+            for (auto i = 0; i < asset::EDT_COUNT; ++i)
+            {
+                if ((m_flags & IDescriptorPool::ECF_FREE_DESCRIPTOR_SET_BIT) && (m_maxDescriptorCount[i] > 0))
+                    _NBL_ALIGNED_FREE(m_generalAllocatorReservedSpace[i]);
+            }
+        }
+
+        // This will return the offset into the pool's descriptor storage. These offsets will be combined
+        // later with base memory addresses to get the actual memory adress where we put the core::smart_refctd_ptr<const IDescriptor>.
+        uint32_t allocateDescriptors(const asset::E_DESCRIPTOR_TYPE type, const uint32_t count)
+        {
+            const uint32_t bytesToAllocate = count * sizeof(void*);
+
+            uint32_t offset;
+            uint32_t invalidAddress;
+            if (m_flags & ECF_FREE_DESCRIPTOR_SET_BIT)
+            {
+                offset = m_generalAllocators[type].alloc_addr(bytesToAllocate, 1u);
+                invalidAddress = core::GeneralpurposeAddressAllocator<uint32_t>::invalid_address;
+            }
+            else
+            {
+                offset = m_linearAllocators[type].alloc_addr(bytesToAllocate, 1u);
+                invalidAddress = core::LinearAddressAllocator<uint32_t>::invalid_address;
+            }
+
+            return (offset == invalidAddress) ? ~0u : offset;
+        }
+
+        void freeDescriptors(const uint32_t count, void* descriptors, const asset::E_DESCRIPTOR_TYPE type)
+        {
+            assert(m_flags & ECF_FREE_DESCRIPTOR_SET_BIT);
+            uint32_t allocatedAddr = static_cast<uint32_t>(reinterpret_cast<uint8_t*>(descriptors) - getDescriptorMemoryBaseAddress(type));
+            m_generalAllocators[type].free_addr(allocatedAddr, count*sizeof(core::smart_refctd_ptr<const asset::IDescriptor>));
+        }
```

Comment:
If you declare these methods pure virtual in `IDescriptorSet` and then move these definitions to `IGPUDescriptorSet` (by making it a friend of the `IDescriptorPool`) as the virtual method implementation, then this becomes the dependency injection we talked about.

Basically the `ICPUDescriptorSet` would "self allocate" while `IGPUDescriptorSet` would allocate from the pool.

Reply:
Resolved.

Reply:
The code for `freeDescriptors` is inlined in `IDescriptorPool::freeDescriptorSets` and I have moved `allocateDescriptors` as a a pure `virtual` method in `IDescriptorSet`.

Reply:
I'm really dumb, sorry I've made you go around in circles.

You're right `IEmulatedDescriptorSet` has no right to exist really, `IDescriptorPool` covers everything

there was no need to move `allocateDescriptors` to IDescriptorSet`

Reply:
Resolved.